compiler/rustc_target/src/abi/mod.rs

```diff
@@ -779,17 +786,29 @@ impl WrappingRange {
         self.end = end;
         self
     }
+
+    /// Returns `true` if `size` completely fills the range.
+    #[inline]
+    pub fn is_full_for(&self, size: Size) -> bool {
+        let max_value = size.unsigned_int_max();
+        debug_assert!(self.start <= max_value && self.end <= max_value);
+        (self.start == 0 && self.end == max_value) || (self.end.wrapping_add(1) == self.start)
```

Comment:
Yeah, looking at the original, I see `start..(end.wrapping_add(1) & mask)`, so `& max_value` is definitely needed.

But also, that condition on its own will account for `self.start == 0 && self.end == max_value` (since `self.end.wrapping_add(1) & max_value` will be `0` i.e. `self.start`), so that part can be removed.

Reply:
Good catch, originally I took the code from https://github.com/rust-lang/rust/blob/434cb437b55d61bcb54a01921de7ac752e6dee13/compiler/rustc_mir/src/interpret/validity.rs#L631
Will fix it.

Reply:
@eddyb I think with the extra `||`, it does not need the `& mask`. So I think the old code was correct. But maybe you find the bit magic version cleaner?

Reply:
Without the masking, the operation is not equivalent to "add 1 *as the unsigned integer type of the given size*", which makes it more confusing for me as to what the intent is.

It's less "bit magic" and more "selecting the right ring to do arithmetic on", it just so happens that `mod 2**n` is implemented as `& (2**n - 1)`.

The removal of the redundant check is just a small bonus - after all, the condition we're looking for is "there is no gap between `end` and `start`, `mod 2**n`".

Reply:
I'd much prefer an actual `mod` and letting the compiler sort this out, that would be so much more readable.^^ The current code is obfuscated, as far as I am concerned. But annoying finite arithmetic makes that tricky too...

Reply:
I would gladly support using `%` if `2**128` could fit in an `u128`.
One idea I was considering is having something like `% Pow2(int_size.bits())`, that does the right thing internally.

Or maybe a wrapper for `u128` that tracks the integer size dynamically and performs all operations according to that size (I suspect miri would like something like this in several places).